Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs

```diff
@@ -161,6 +208,76 @@ protected override void FinishSetup(InputDeviceBuilder builder)
             paddle2 = builder.GetControl<ButtonControl>(this, "paddle2");
             paddle3 = builder.GetControl<ButtonControl>(this, "paddle3");
             paddle4 = builder.GetControl<ButtonControl>(this, "paddle4");
+        }
+
+        public ulong gamepadId
+        {
+            get
+            {
+                UpdatePadSettings();
+                return m_GamepadId;
+            }
+        }
+
+        public int xboxOneUserId
+        {
+            get
+            {
+                UpdatePadSettings();
+                return m_XboxOneUserId;
+            }
+        }
+
+        public new static List<XboxOneGamepad> all
```

Comment:
No lists. Use array for s_Devices and ReadOnlyArray as return type here.

Reply:
Really? The problem is it doesn't know how many there might be. Also returning an array from a list just allocates more memory and creates garbage. We do try to minimize that on the consoles as the garbage collect is more expensive on consoles. However I'll sort out this change.

Reply:
You can use ArrayHelpers to allocate with capacity (same thing that list does internally). E.g. with a capacity of 10. With a list you get two heap allocations. For the price you pay for the List header object, you can easily pay for 5 extra entries in your array. And List.Append() is not guarantee at all to not create garbage. Internally, it just allocates with capacity as well (same thing you can do with ArrayHelpers).

Reply:
Also, no returning mutable internal structures from public APIs.